src/wasm-lib/kcl/src/execution/memory.rs

```diff
@@ -134,8 +154,87 @@
 //! - Therefore, an active env must either be on the call stack, have snapshots, or be a root env. This
 //!   is however a conservative approximation since snapshots may exist even if there are no live
 //!   references to an env.
-
-use std::{collections::HashMap, fmt};
+//!
+//! ## Concurrency and thread-safety
+//!
+//! `ProgramMemory` is a global singleton (technically one per program execution, if we handled multiple
+//! projects in a single interpreter process we'd need multiple `ProgramMemory`s, but that is currently
+//! not possible). `ProgramMemory` could be moved between threads, but there shouldn't be any need
+//! to do so. It can safely be referenced and accessed from multiple threads, but there are rules for
+//! doing so.
+//!
+//! `ProgramMemory` is mostly accessed via a `Stack` object, avoid accessing `ProgramMemory` directly
+//! where possible. `Stack`s can safely be moved to other threads and can access `ProgramMemory`
+//! from a different thread. There can be multiple `Stack`s on different threads or the same thread
+//! (either operating sequentially or using async tasks).
+//!
+//! The key requirement for users is that names in from a `Stack` should never be exposed until the
```

Comment:
Is this trying to say...?
```suggestion
//! The key requirement for users is that names from a `Stack` should never be exposed until the
```